### PR TITLE
(fix): add support for TLS certificates pinning

### PR DESCRIPTION
**Pull request by jaeopt**
## Summary
Some applications require support for TLS pinning for extra security. This fix allows opt-in for that feature.
 
- add opt-in settings (off by default - most applications do not need this)
- embed public root certificates and verify if certs not compromised when pinning is enabled.
- this feature is for iOS only (not supported for tvOS)

## Test plan
- Add new unit tests for 4 combinations: {(disabled, good), (enabled, good), (disabled, bad), (enabled, bad)}
- Add FSC test suites

## Issues
- OASIS-5056

### CODE REVIEW

File: OptimizelySDKEventDispatcher/OptimizelySDKEventDispatcher/OPTLYEventDispatcher.h [hunk 1]
```diff
@@ -57,6 +57,9 @@ typedef void (^OPTLYEventDispatcherResponse)(NSData * _Nullable data, NSURLRespo
 /// Max number of events to store before overwriting older events (value must be greater than 1)
 @property (nonatomic, assign, readonly) NSInteger maxNumberOfEventsToSave;
 
+/// Enable TLS pinning for extra security (do not turn this on unless you really need extra TLS security
```
Comment (yasirfolio3) [inline OptimizelySDKEventDispatcher/OptimizelySDKEventDispatcher/OPTLYEventDispatcher.h]: (nit) Missing closing bracket.

File: OptimizelySDKEventDispatcher/OptimizelySDKEventDispatcher/OPTLYEventDispatcherBuilder.h [hunk 2]
```diff
@@ -44,6 +44,8 @@ typedef void (^OPTLYEventDispatcherBuilderBlock)(OPTLYEventDispatcherBuilder * _
 @property (nonatomic, assign, readwrite) NSInteger eventDispatcherDispatchInterval;
 /// Max number of events to store before overwriting older events (value must be greater than 1)
 @property (nonatomic, assign) NSInteger maxNumberOfEventsToSave;
+/// Enable TLS pinning for extra security (do not turn this on unless you really need extra TLS security
```
Comment (yasirfolio3) [inline OptimizelySDKEventDispatcher/OptimizelySDKEventDispatcher/OPTLYEventDispatcherBuilder.h]: (nit) missing closing bracket.

File: OptimizelySDKShared/OptimizelySDKShared/OPTLYManagerBuilder.h [hunk 3]
```diff
@@ -27,6 +27,8 @@ typedef void (^OPTLYManagerBuilderBlock)(OPTLYManagerBuilder * _Nullable builder
 @property (nonatomic, readwrite, strong, nullable) NSData *datafile;
 /// The dispatch interval for the event dispatcher.
 @property (nonatomic, readwrite) NSTimeInterval eventDispatchInterval;
+/// Enable TLS pinning for extra security (do not turn this on unless you really need extra TLS security
```
Comment (yasirfolio3) [inline OptimizelySDKShared/OptimizelySDKShared/OPTLYManagerBuilder.h]: (nit) bracket missing.

File: OptimizelySDKiOS/OptimizelySDKiOS/OPTLYManager.m [hunk 4]
```diff
@@ -92,13 +92,18 @@ - (instancetype)initWithBuilder:(OPTLYManagerBuilder *)builder {
         
         self.datafileConfig = [[OPTLYDatafileConfig alloc] initWithProjectId:self.projectId withSDKKey:self.sdkKey];
         
+        // --- TLS certificates pinning (disabled by default) ---
+        // TLS pinnig is supported for iOS only (not tvOS)
```
Comment (yasirfolio3) [inline OptimizelySDKiOS/OptimizelySDKiOS/OPTLYManager.m]: (nit) change pinnig to pinning.

File: OptimizelySDKiOS/OptimizelySDKiOSTests/OptimizelySDKiOSTLSPinningOptInTests.m [hunk 5]
```diff
@@ -0,0 +1,129 @@
+/****************************************************************************
+ * Copyright 2016-2018, Optimizely, Inc. and contributors                   *
```
Comment (yasirfolio3) [inline OptimizelySDKiOS/OptimizelySDKiOSTests/OptimizelySDKiOSTLSPinningOptInTests.m]: (nit) Please update the years.

File: OptimizelySDKiOS/OptimizelySDKiOSTests/OptimizelySDKiOSTLSPinningOptInTests.m [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        // pinning is off by default
+    }]];
+    
+    XCTAssert(pinningDisabledCount > 0);
+    XCTAssert(pinningEnabledCount == 0);
+    
+    [self swizzleCreateSessionMethods];
+}
+
+- (void)testTLSPinngOptOut {
```
Comment (yasirfolio3) [inline OptimizelySDKiOS/OptimizelySDKiOSTests/OptimizelySDKiOSTLSPinningOptInTests.m]: (nit) change Pinng to Pinning. Same for all test names.

File: OptimizelySDKDatafileManager/OptimizelySDKDatafileManager/OPTLYDatafileManagerBuilder.h [hunk 7]
```diff
@@ -35,6 +35,8 @@ typedef void (^OPTLYDatafileManagerBuilderBlock)(OPTLYDatafileManagerBuilder * _
  * The default time interval is 0. This means that the datafile manager will NOT regularly poll for a new datafile during the app session.
  */
 @property (nonatomic, readwrite) NSTimeInterval datafileFetchInterval;
+/// Enable TLS pinning for extra security (do not turn this on unless you really need extra TLS security
```
Comment (yasirfolio3) [inline OptimizelySDKDatafileManager/OptimizelySDKDatafileManager/OPTLYDatafileManagerBuilder.h]: (nit) Missing closing bracket.

File: OptimizelySDKDatafileManager/OptimizelySDKDatafileManager/OPTLYDatafileManager.h [hunk 8]
```diff
@@ -32,6 +32,8 @@
 
 /// The time interval to regularly fetch the datafile.
 @property (nonatomic, readonly) NSTimeInterval datafileFetchInterval;
+/// Enable TLS pinning for extra security (do not turn this on unless you really need extra TLS security
```
Comment (yasirfolio3) [inline OptimizelySDKDatafileManager/OptimizelySDKDatafileManager/OPTLYDatafileManager.h]: (nit) Missing closing bracket.

File: OptimizelySDKCore/OptimizelySDKCoreTests/TLSPinningTests.m [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    }
+    
+    [self waitForExpectationsWithTimeout:5.0 handler:^(NSError *error) {
+        if (error) {
+            NSLog(@"Timeout error for testGETSuccess: %@", error);
+        }
+    }];
+}
+
+// test with wrong root certficates
```
Comment (yasirfolio3) [inline OptimizelySDKCore/OptimizelySDKCoreTests/TLSPinningTests.m]: (nit): change certficates to certificates.

File: OptimizelySDKCore/OptimizelySDKCoreTests/TLSPinningTests.m [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+// - try to connect other urls using different root certificates
+// - they're all valid certs but root cert will be rejected by comparing with pinned ones
+// - when pinning is disabled (default), all connections to any ursl must be successful
```
Comment (yasirfolio3) [inline OptimizelySDKCore/OptimizelySDKCoreTests/TLSPinningTests.m]: (nit): change ursl to urls.

File: OptimizelySDKCore/OptimizelySDKCoreTests/TLSPinningTests.m [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    self.testURL = [NSURL URLWithString:kTestURLString];
+}
+
+- (void)tearDown {
+    [OHHTTPStubs removeAllStubs];
+    self.testURL = nil;
+    [super tearDown];
+}
+
+#pragma mark - TLS Certficates Pinning Tests
```
Comment (yasirfolio3) [inline OptimizelySDKCore/OptimizelySDKCoreTests/TLSPinningTests.m]: (nit): change Certficates to Certificates.

File: OptimizelySDKCore/OptimizelySDKCoreTests/TLSPinningTests.m [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+// test with correct root certficates
```
Comment (yasirfolio3) [inline OptimizelySDKCore/OptimizelySDKCoreTests/TLSPinningTests.m]: (nit): Certificates.

File: OptimizelySDKCore/OptimizelySDKCoreTests/TLSPinningTests.m [hunk 13]
```diff
@@ -0,0 +1,164 @@
+/****************************************************************************
+ * Copyright 2016-2017, Optimizely, Inc. and contributors                   *
```
Comment (yasirfolio3) [inline OptimizelySDKCore/OptimizelySDKCoreTests/TLSPinningTests.m]: (nit): update years.

**Review by yasirfolio3 — CHANGES_REQUESTED** [review]
LGTM, just few nits.

**Comment by jaeopt** [comment]
SSL pinning support is put on hold. We'll review the issue and solution again later.

**Pull request closed without merging**